// @ctrl/ngx-csv/esm5/csv.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { buildURI } from './util';
var CsvDirective = /** @class */ (function () {
    function CsvDirective(sanitizer) {
        this.sanitizer = sanitizer;
        /** the body of the csv */
        this.data = [];
        /** Set the seperator between values */
        this.delimiter = ',';
        /** adds a Byte order mark to setup the csv as UTF-8  */
        this.uFEFF = true;
        /** filename */
        this.download = 'data.csv';
        this.target = '_blank';
    }
    Object.defineProperty(CsvDirective.prototype, "filename", {
        /** Set the filename of the csv. Default is `data.csv` */
        set: function (a) {
            this.download = a;
        },
        enumerable: true,
        configurable: true
    });
    CsvDirective.prototype.ngOnChanges = function () {
        this.href = this.sanitizer.bypassSecurityTrustResourceUrl(buildURI(this.data, this.uFEFF, this.headers, this.delimiter));
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "data", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], CsvDirective.prototype, "headers", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "delimiter", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], CsvDirective.prototype, "filename", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "uFEFF", void 0);
    tslib_1.__decorate([
        HostBinding(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "href", void 0);
    tslib_1.__decorate([
        HostBinding(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "download", void 0);
    tslib_1.__decorate([
        Input(), HostBinding(),
        tslib_1.__metadata("design:type", Object)
    ], CsvDirective.prototype, "target", void 0);
    CsvDirective = tslib_1.__decorate([
        Directive({ selector: '[csvLink]' }),
        tslib_1.__metadata("design:paramtypes", [DomSanitizer])
    ], CsvDirective);
    return CsvDirective;
}());
export { CsvDirective };
