// ngx-filemanager-core/esm2015/FileSystemProvider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function FileSystemProvider() { }
if (false) {
    /**
     * @param {?} path
     * @return {?}
     */
    FileSystemProvider.prototype.List = function (path) { };
    /**
     * @param {?} newPath
     * @param {?=} disableNoClobber
     * @return {?}
     */
    FileSystemProvider.prototype.CreateFolder = function (newPath, disableNoClobber) { };
    /**
     * @param {?} singleFileName
     * @param {?} newPath
     * @return {?}
     */
    FileSystemProvider.prototype.Copy = function (singleFileName, newPath) { };
    /**
     * @param {?} item
     * @param {?} newPath
     * @return {?}
     */
    FileSystemProvider.prototype.Move = function (item, newPath) { };
    /**
     * @param {?} item
     * @param {?} newItemPath
     * @return {?}
     */
    FileSystemProvider.prototype.Rename = function (item, newItemPath) { };
    /**
     * @param {?} item
     * @param {?} content
     * @return {?}
     */
    FileSystemProvider.prototype.Edit = function (item, content) { };
    /**
     * @param {?} item
     * @return {?}
     */
    FileSystemProvider.prototype.Getcontent = function (item) { };
    /**
     * @param {?} item
     * @param {?} role
     * @param {?} entity
     * @param {?=} recursive
     * @return {?}
     */
    FileSystemProvider.prototype.SetPermissions = function (item, role, entity, recursive) { };
    /**
     * @param {?} items
     * @param {?} newPath
     * @return {?}
     */
    FileSystemProvider.prototype.MoveMultiple = function (items, newPath) { };
    /**
     * @param {?} items
     * @param {?} newPath
     * @return {?}
     */
    FileSystemProvider.prototype.CopyMultiple = function (items, newPath) { };
    /**
     * @param {?} items
     * @param {?} role
     * @param {?} entity
     * @param {?=} recursive
     * @return {?}
     */
    FileSystemProvider.prototype.SetPermissionsMultiple = function (items, role, entity, recursive) { };
    /**
     * @param {?} items
     * @param {?} permissionsObj
     * @param {?=} recursive
     * @return {?}
     */
    FileSystemProvider.prototype.SetPermissionsObjectMultiple = function (items, permissionsObj, recursive) { };
    /**
     * @param {?} items
     * @return {?}
     */
    FileSystemProvider.prototype.Remove = function (items) { };
    /**
     * @param {?} currentPath
     * @return {?}
     */
    FileSystemProvider.prototype.GetUploadApiUrl = function (currentPath) { };
    /**
     * @param {?} file
     * @return {?}
     */
    FileSystemProvider.prototype.CreateDownloadLink = function (file) { };
    /**
     * @param {?} item
     * @return {?}
     */
    FileSystemProvider.prototype.GetSingle = function (item) { };
    /**
     * @param {?} item
     * @return {?}
     */
    FileSystemProvider.prototype.Upload = function (item) { };
    /**
     * @return {?}
     */
    FileSystemProvider.prototype.CloneProvider = function () { };
}
